include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +393,277 @@ class BidirectionalNode
   NeighborContainer Predecessors;
 };
 
+template<typename T>
+concept IsBidirectionalNode = requires {
+  T::is_bidirectional_node;
+  typename llvm::Inverse<T *>;
+};
+
+namespace detail {
+/// The parameters deciding specifics of the base type `NewNode` extends are
+/// non-trivial. That's why those decisiion were wrapped inside this struct
+/// to minimize clutter.
+///
+/// \note At this step `TheNode` has not been declared yet, thus we accept a
+///       template parameter that has the same signature as `TheNode` that
+///       will be declared later. This allows us to use it as if it was
+///       declared, provided that only the real `TheNode` is used as this
+///       argument.
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename, typename, bool, size_t, typename, size_t, bool>
+         class TheNode,
+         typename FinalType,
+         size_t ParentSmallSize,
+         bool ParentHasEntryNode>
+struct NewNodeBaseTCalc {
+  static constexpr bool
+    NoDerivation = std::is_same_v<FinalType, std::false_type>;
+  using NodeType = TheNode<Node,
+                           EdgeLabel,
+                           HasParent,
+                           SmallSize,
+                           FinalType,
+                           ParentSmallSize,
+                           ParentHasEntryNode>;
+  using DerivedType = std::conditional_t<NoDerivation, NodeType, FinalType>;
+  using GenericGraph = GenericGraph<DerivedType,
+                                    ParentSmallSize,
+                                    ParentHasEntryNode>;
+  using ParentType = Parent<GenericGraph, Node>;
+  using Result = std::conditional_t<HasParent, ParentType, Node>;
+};
+} // namespace detail
+
+/// A node type with support for non-trivial edge types.
```

Comment:
Can you please document the assumptions?
For instance can I have multiple outgoing edges going to the same successor (with or without the same label)?

Reply:
Yes, we decided to mark "duplicate" edges UB, and it definitely need to be documented. Sorry.

Reply:
BTW, on this note, maybe we should go a bit further and use an assert, say in `AddSuccessor` we try to `findSuccessor` before actually adding it?
In optimized builds the check is to be optimized anyway, but it can be a nice safeguard while debugging. Moreover, assert is always safer that a comment.

Reply:
Sure, then, if it's *very* expensive we can `#ifdef` it behind `EXPENSIVE_CHECKS`.

Reply:
It's expensive is a sense that it loops over containers. Rather "non-trivial" than expensive.

Reply:
What I want to confirm is that `revng_check`s are properly removed in `optimized` builds, right? Or maybe even just removing those checks from the `release` builds is good enough.
Either way, `ifdef` is definitely an overkill here ~~(and is ugly)~~.

Reply:
Nope, `optimized` retains asserts. `release` drops them.
I think that if it becomes a bottleneck in some use case we'll notice it and act on it later on.
We have some checks under `if (VerifyLogger.isEnabled())` so it can be enabled at run-time.

Reply:
So then I'm adding `revng_assert(findSuccessor(Arg0) == Successors.end())` to the `addSuccessor` and friends for now. And if something comes up, we can deal with it later, right?